waiter/src/waiter/core.clj

```diff
@@ -949,14 +955,19 @@
                                       (fn service-description->service-id [service-description]
                                         (sd/service-description->service-id service-id-prefix service-description)))
    :service-id->idle-timeout (pc/fnk [[:settings [:token-config token-defaults]]
-                                      service-id->service-description-fn service-id->source-tokens-entries-fn
+                                      [:state service-description-builder]
+                                      service-id->service-description-fn service-id->references-fn
                                       token->token-hash token->token-metadata]
-                               (fn service-id->idle-timeout [service-id]
-                                 (sd/service-id->idle-timeout
-                                   service-id->service-description-fn service-id->source-tokens-entries-fn
-                                   token->token-hash token->token-metadata token-defaults service-id)))
+                               (let [context {:token->token-hash token->token-hash}
+                                     reference-type->stale-fn (sd/retrieve-reference-type->stale-fn service-description-builder context)]
+                                 (fn service-id->idle-timeout [service-id]
+                                   (sd/service-id->idle-timeout
+                                     service-id->service-description-fn service-id->references-fn token->token-metadata
+                                     reference-type->stale-fn token-defaults service-id))))
    :service-id->password-fn (pc/fnk [[:scheduler service-id->password-fn*]]
                               service-id->password-fn*)
+   :service-id->references-fn (pc/fnk [[:state kv-store]]
+                                (partial sd/service-id->references kv-store))
```

Comment:
Can we avoid `partial`?  It's quite slow and this is in the critical path.

Reply:
I was curious about this and dug a bit further.

https://stuartsierra.com/2015/06/01/clojure-donts-optional-arguments-with-varargs

Were you aware that calling varargs functions always allocates a new sequence?  

https://gist.github.com/halgari/5d7780e6f068ae217bb2

We use these a lot and I wonder if we can improve performance and reduce GC pressure by avoiding these.

Reply:
Separate PR though